### PR TITLE
Add notification fetch status to app.-status and don't create request while in progress

### PR DESCRIPTION
**Pull request by daehn**
# What's in this PR?

* We now query if we are currently fetching the notification stream as response to receiving a push notification before creating requests in the client message transcoder.
* Fix `LinkPreviewUploadRequestStrategy` generating events during sync.

### CODE REVIEW

File: Source/Transcoders/ClientMessageTranscoder.swift [hunk 1]
```diff
@@ -49,6 +49,10 @@ public class ClientMessageTranscoder: AbstractRequestStrategy {
     }
     
     public override func nextRequestIfAllowed() -> ZMTransportRequest? {
+        // Don't create requests while we are still fetching the notification stream in the background.
+        // Otherwise we risk already sending out delivery receipts for received messages when we have to fetch
+        // multiple pages of the stream (in case we have been offline for a while).
+        guard applicationStatus?.notificationFetchStatus.status == .done else { return nil }
```
Comment (daehn) [inline Source/Transcoders/ClientMessageTranscoder.swift]: Currently this transcoder should be the only one doing requests at this point, but we might want to move this into the prerequisites as well.

Comment (daehn) [inline Source/Transcoders/ClientMessageTranscoder.swift]: What do you think @typfel?

Comment (typfel) [inline Source/Transcoders/ClientMessageTranscoder.swift]: I think we can go with this until we need it somewhere else, shouldn't be big effort to re-factor it later.

Comment (daehn) [inline Source/Transcoders/ClientMessageTranscoder.swift]: @typfel to late, already done 😛

File: Source/Transcoders/ClientMessageTranscoder.swift [hunk 2]
```diff
@@ -49,6 +49,10 @@ public class ClientMessageTranscoder: AbstractRequestStrategy {
     }
     
     public override func nextRequestIfAllowed() -> ZMTransportRequest? {
+        // Don't create requests while we are still fetching the notification stream in the background.
+        // Otherwise we risk already sending out delivery receipts for received messages when we have to fetch
+        // multiple pages of the stream (in case we have been offline for a while).
+        guard applicationStatus?.notificationFetchStatus == .done else { return nil }
```
Comment (daehn) [inline Source/Transcoders/ClientMessageTranscoder.swift]: Seoul be removed

**Comment by typfel** [comment]
Have you also checked the strategies in SE?

**Comment by daehn** [comment]
@typfel doing that as we speak

**Comment by daehn** [comment]
@typfel updated https://github.com/wireapp/wire-ios-sync-engine/pull/418

File: Source/Transcoders/LinkPreviewUploadRequestStrategy.swift [hunk 3]
```diff
@@ -31,19 +31,16 @@ fileprivate extension ZMClientMessage {
...
-    /// Auth status to know whether we can make requests
-    fileprivate var clientRegistrationDelegate: ClientRegistrationDelegate
-
     /// Upstream sync
     fileprivate var upstreamSync: ZMUpstreamModifiedObjectSync!
 
-    public init(managedObjectContext: NSManagedObjectContext, clientRegistrationDelegate: ClientRegistrationDelegate) {
-        self.clientRegistrationDelegate = clientRegistrationDelegate
-        super.init(managedObjectContext: managedObjectContext)
+    public override init(withManagedObjectContext managedObjectContext: NSManagedObjectContext, applicationStatus: ApplicationStatus) {
```
Comment (daehn) [inline Source/Transcoders/LinkPreviewUploadRequestStrategy.swift]: API breaker

**Pull request merged**